pilot/pkg/serviceregistry/kube/controller/pod.go

```diff
@@ -82,46 +90,75 @@ func (pc *PodCache) onEvent(curr interface{}, ev model.Event) error {
 
 	ip := pod.Status.PodIP
 
+	// trigger the eds update when the pods are ready, or become unready
+	edsUpdateRequired := false
+
 	// PodIP will be empty when pod is just created, but before the IP is assigned
 	// via UpdateStatus.
 	if len(ip) > 0 {
 		key := kube.KeyFunc(pod.Name, pod.Namespace)
 		switch ev {
 		case model.EventAdd:
 			switch pod.Status.Phase {
-			case v1.PodPending, v1.PodRunning:
+			case v1.PodPending:
+				if key != pc.podsByIP[ip] {
+					// add to cache if the pod is pending
+					// This is for keeping the old logic
+					pc.update(ip, key)
+				}
+			case v1.PodRunning:
+				if isPodReady(pod) {
+					edsUpdateRequired = true
+				}
 				if key != pc.podsByIP[ip] {
-					// add to cache if the pod is running or pending
 					pc.update(ip, key)
 				}
 			}
 		case model.EventUpdate:
 			if pod.DeletionTimestamp != nil {
+				edsUpdateRequired = true
+				ev = model.EventDelete
 				// delete only if this pod was in the cache
 				if pc.podsByIP[ip] == key {
 					pc.deleteIP(ip)
 				}
-				ev = model.EventDelete
 			} else {
 				switch pod.Status.Phase {
-				case v1.PodPending, v1.PodRunning:
+				case v1.PodPending:
 					if key != pc.podsByIP[ip] {
-						// add to cache if the pod is running or pending
 						pc.update(ip, key)
 					}
-
+				case v1.PodRunning:
+					if isPodReady(pod) {
+						edsUpdateRequired = true
+						if key != pc.podsByIP[ip] {
+							pc.update(ip, key)
+						}
+					}
+					// When running pods become unready, we also need to trigger eds update
+					if key == pc.podsByIP[ip] && !isPodReady(pod) {
+						pc.deleteIP(ip)
```

Comment:
This may fix ServiceEntry selection but I am fairly sure this will completely break standard pods. We need the pod information before it becomes ready, so we can send it the initial metadata

Reply:
> I am fairly sure this will completely break standard pods
In what way? Seems like the test cases are passed assuming Istio has the tests for that part...From my local testing, I can see the endpoints are being correctly sent to sidecar proxy, both the k8s fqdn and service entry. 

The big difference of this change is that eds update is not triggered every time when there is an event. After changes, Only pods that have passed readiness check will be sent to sidecar. We also need to take care when the pods become unready.

Reply:
> The big difference of this change is that eds update is not triggered every time when there is an event. After changes, Only pods that have passed readiness check will be sent to sidecar.

We actually *do* want to fetch the pod before it becomes ready. There are two cases for fetching the pod: (1) we are sending config to that pod, and need info about it. (2) we are using it to send as an endpoint for other pods.

This PR seems to improve case (2), but it is at the expense of case (1). The only reason the tests pass, I suspect, is because we have a fallback in place to use the node metadata from the XDS connection to handle cases where their is propagation delay in k8s, but this is intended for fallback only not standard case

Reply:
> The only reason the tests pass, I suspect, is because we have a fallback in place to use the node metadata from the XDS connection to handle cases where their is propagation delay in k8s, but this is intended for fallback only not standard case

I guess this also explains why the pod itself becomes ready in my local testing. So basically we need to separate the eds push with these two cases. They are now mixed together and is causing the issue with case 2). Actually the No.2 case has much bigger impact if Istio has a fallback for case 1. 

Need a way to figure out these two separate pushes...